spring-pulsar-docs/src/main/antora/modules/ROOT/pages/reference/pulsar/transactions.adoc

```diff
@@ -0,0 +1,222 @@
+<4> Listener method sends transformed message with transactional template which enlists send operation in active transaction
+<5> Listener container auto-acks message and enlists ack operation in active transaction
+<6> Listener container (via `TransactionTemplate`) commits transaction
+
+If you are not using `@PulsarListener` and instead using listener containers directly, the same transaction support is provided as described above.
+Remember, the `@PulsarListener` is just a convenience to register a Java method as the listener container message listener.
+
+=== Transactions with Record Listeners
+The above example uses a record listener.
+When using a record listener, a new transaction is created on every listener method invocation which equates to a transaction per message.
+
+NOTE: Because the transaction boundary is per message and each message acknowledgement is enlisted in each transaction, batch ack mode  can not be used with transactional record listeners.
+
+=== Transactions with Batch Listeners
+When using a batch listener, a new transaction is created on every listener method invocation which equates to a transaction per batch of messages.
+
+NOTE: Due to complexity, transactional batch listeners do not currently support custom error handlers.
+
+=== Configuration
+
+==== Listener container factory
+The following transaction settings are available directly on the `PulsarContainerProperties` used by the `ConcurrentPulsarListenerContainerFactory` when creating listener containers.
+These settings affect all listener containers, including the ones used by `@PulsarListener`.
+
+* `enabled` - whether the container supports transactions (default `false`)
+
+* `required` - whether the container requires transactions (default `false`)
+
+* `timeout` - duration of the transaction timeout (default `null`)
+
+* `transactionDefinition` - a blueprint transaction definition with properties that will be copied to the container's transaction template (default `null`)
+
+* `transactionManager` - the transaction manager used to start transactions
+
+When not using Spring Boot, you can adjust these settings on the container factory that you provide.
+However, when using Spring Boot, the container factory is auto-configured.
+In this case you can register a `ConcurrentPulsarListenerContainerFactoryCustomizer` bean to access and customize the container properties.
+The following example shows how to set the timeout on the container factory:
+
+[source, java]
+----
+@Bean
+ConcurrentPulsarListenerContainerFactoryCustomizer containerCustomizer() {
+    return (containerFactory) -> containerFactory.getContainerProperties().transactions().setTimeout(Duration.ofSeconds(45));
+}
+----
+
+==== `@PulsarListener`
+By default, each listener respects the transactional settings of its corresponding listener container factory.
+However, the user can set the `transactional` attribute on each `@PulsarListener` to override the container factory setting as follows:
```

Comment:
i'm a little confused, it looks like on line 154 our listeners can override the transaction attributes of the factory.   BUt the points below say that we'll get an exception if we disable it.   But I could be misreading it.

Reply:
The points below are to outline the rules of setting `@PulsarListener(transactional = "")`. 

The only thing we can override really is "enabled" and its conditional on the following rules: 

- If `transactional` not specified - listener inherits settings from container factory (as if the attribute does not exist)

- if `transactional = false` and container factory does not have `required = true` then it will disable txn for listener

- if `transactional = false` and container factory does have `required = true` then txn are always required and this will lead to exception

- if `transactional = true` but container factory does have `enabled = true` then txn are not even available for the container factory, much less for individual listeners (we log this fact if it happens)

I am totally up for suggestions to make this less confusing. It is a bit confusing.

Reply:
After rereading the rules again it makes sense.    I was getting required and enabled confused.